test/functional/example_y2038.py

```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env python3
+# Copyright (c) 2020 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test covering the Year 2038 problem.
+
+https://en.wikipedia.org/wiki/Year_2038_problem
+"""
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import (
+    assert_equal,
+)
+
+EPOCH_2038_BEFORE = 2147440306 # (GMT): Monday, January 18, 2038 3:11:46 PM
+EPOCH_2038_AFTER = 2147613106 # (GMT): Wednesday, January 20, 2038 3:11:46 PM
+EPOCH_2106 = 4293404400000
+
+
+class Y2038EpochTest(BitcoinTestFramework):
+
+    def set_test_params(self):
+        self.num_nodes = 2
+
+    def test_y2038(self):
+
+        for epoch in [EPOCH_2038_BEFORE, EPOCH_2038_AFTER, EPOCH_2106]:
+            # mocks the time to the epoch
+            self.nodes[0].setmocktime(epoch)
+            self.nodes[1].setmocktime(epoch)
+
+            # Generates a blocks
+            self.nodes[0].generate(nblocks=1)
```

Comment:
Note that the expected behaviour for AD 2106 and beyond is for all blocks to be invalid... This should fail.

Reply:
yeah thats because there isnt not enough bits are difined at specific parts of the code. as example: https://github.com/bitcoin/bitcoin/blob/4a540683ec40393d6369da1a9e02e45614db936d/src/primitives/block.h#L27

Reply:
Changing the type there is a hardfork. Also not a good approach (makes more sense to just allow it to wrap around).

Defining a hardfork goes beyond writing tests. A test should simply confirm such blocks are invalid.

Reply:
> 
> 
> Changing the type there is a hardfork. Also not a good approach (makes more sense to just allow it to wrap around).
> 
> Defining a hardfork goes beyond writing tests. A test should simply confirm such blocks are invalid.

Well a approach could be that you first implement this type change in a patch and distribute it but just not activate it yet. You just set a block-id when this update should be active. That its enough time to let the hole network been updated. You basicly keep using for the old blocks the old uint32_t type. But new blocks will be uint_64 after that.

Reply:
Got it. I'm going to change the test to reflect that. I also saw it failed at one of our CI checks (https://cirrus-ci.com/task/6103755461492736?command=ci#L5253), but not on the others https://cirrus-ci.com/task/4626011833761792?command=ci#L3610

```
 node0 stderr miner.cpp:130:21: runtime error: implicit conversion from type 'int64_t' (aka 'long') of value 4293404400000 (64-bit, signed) to type 'uint32_t' (aka 'unsigned int') changed the value to 2732071296 (32-bit, unsigned)
```